CONFETI-28 feat: SQS 초기 세팅

jher235 opened the pull request:
## 🔊 Summaries
<!-- 본인이 한 작업을 요약해주세요. -->

- Apple Music API에 대한 의존성을 줄이고 응답 성능을 개선하기 위해, 요청된 데이터를 서비스 DB에 이중화하여 관리하는 로직을 도입하고자 함.
- 클라이언트가 Artist, Song 데이터를 요청할 경우 우선적으로 서비스 DB를 조회하여 데이터가 존재하지 않을 경우에만 Apple Music API를 호출하여 정보를 가져온 후 사용자에게 반환함. 이때 API를 통해 새로 가져온 데이터는 비동기적으로 서비스 DB에 저장하며, 데이터 저장 로직의 성공 여부가 사용자에게 영향을 주지 않아야 함. 즉, Apple Music API에서 데이터를 잘 가져왔다면 사용자는 해당 데이터의 저장 성공 여부와는 관계없이 블로킹되지 않고 응답을 성공적으로 받아야 함.
- 단순히 별도의 스레드를 할당하여 처리하는 방식 대신, 향후 확장성을 고려하여 이벤트 기반 아키텍처를 도입했으며, 비용이 효율적인 AWS SQS를 메시지 큐로 선택함.
- DLQ 의 경우 Maximum receives 를 3으로 둠. 보관 기간은 14일.

### SQS Config
- 각 SqsListener 설정값은 스레드 풀을 제외하고 모두 비용 절감 및 효율성에 초점을 둠
- 한 번의 폴링 요청으로 수신할 수 있는 최대 메시지 수는 AWS SDK의 최대값인 10
- 폴링의 경우 최대 초인 20초로 설정.
- 메세지 삭제의 경우 성공 시 개별적으로 진행됨 
- ~~최대 스레드 limit은 2로 폴링하는 스레드는 2개임~~
- ~~comsumer 스레드 풀의 현재 corepool은 20개. 2 (폴링 스레드 ) * 10 (한번에 수신하는 최대 메세지 수) 으로 계산했으며 DB 커넥션 풀 수보다 크지만, 작업 이후 message 삭제 Acknowledge 하는 부분 로직까지 수행하므로 조금 넉넉하게 설정함~~

### SqsStrategy
- 헤더의 confetiEventType를 통해 이벤트 타입을 관리함
- 메시지 추적을 위해 traceId와 이벤트 타입을 속성(Attribute)에 포함함
- ~~@SqsListener를 사용하여 큐의 메시지를 수신하며 eventHandlers 를 통해 이벤트별 핸들러 메서드를 호출하는 구조~~
  - 별도의 스케줄러를 사용해서 수신하도록 변경함. 기존 @SqsListener 로는 한번에 가져오는 메세지 수가 polling 하는 스레드 수 보다 클 수 없었음. 그러나 우리가 원했던 구조는 polling 스레드와 consume 하는 스레드를 별도로 두고 사용하는 것이었으므로 이를 세세하게 컨트롤하고자 직접 스케줄러를 통해 polling 하도록 변경함. polling 스레드는 단일 스레드.
- SQS 에 메세지를 provide 하는 로직은 실패 시 retry 하지 않고 notify를 줌. Artist나 Song을 생성하는 이벤트 등록이 실패하더라도 치명적이지 않다고 판단했기 때문
- 각 이벤트 핸들러는 데이터베이스에 데이터를 생성하기 전, 해당 데이터의 존재 여부를 먼저 확인하여 멱등성을 보장함
- 현재는 SQS 큐를 단일 큐로 사용함. 기본적으로 롱폴링으로 20초를 설정하더라도 (60/20) * 60 * 24 * 30 = 129600 으로 단순 폴링 요청 수만 계산하더라도 13만에 가까운 호출이 발생함. 따라서 현재는 이벤트마다 큐를 두는 것이 아닌 객체 타입과 핸들러 맵을 사용하여 이벤트별 핸들링을 진행함.

### Event
- Event 및 EventHandler 를 추상화하여 각각의 EventHandler 가 알맞은 Event를 처리할 수 있도록 구성함. 
  - ~~다만 Event에서 반환하는 TypeId와 EventHandler에서 반환하는 TypeId 의 코드가 반환하는 결과는 같지만, 공통적으로 추출해서 작성하진 않았기 때문에 이 부분을 공통적으로 추출해서 관리할지 고민중~~
  - 위 부분은 Event 는 자신이 어떤 TypeId로 처리될 지를 알 필요가 없으며, EventHandler가 이를 결정하는 것이 더 유연한 구조라고 판단 판단되어 EventHandler 에서 EventType을 결정하도록 함. 물론 이전에 논의했던 대로 인터페이스에서 디폴트 메서드를 사용하는 부분이 인터페이스의 역할에서 벗어나는 것처럼 느껴져 좋지 않을 수 있지만, 필드가 필요하지 않은 현재 구조에서 모든 구현체들이 공통적으로 처리할 수 있는 부분을 default 메서드로 처리할 수 있다면 코드 중복을 줄일 수 있다고 판단함. 이렇게 할 경우 필요 시 해당 EventHandler의 타입 아이디 반환 메서드만 오버라이드하여 사용하면되기 때문에 Event와 싱크가 맞지 않을 수 있는 위험도도 없어짐.
- 추후 이벤트가 추가될 경우 Event 와 EvenHandler 만 구현하면 되고 SqsStrategy 까지 변경이 전파되지 않음. 생성자에서 EventHandler  빈 목록을 주입 받아서 알맞은 처리가 가능.

### Resolver
- SQS, Kafka 등의 MessageBroker 마다  MessageBrokerStrategy 를 구현해야 함.
- 현재는 SQS 내에서도 목적에 맞게 큐를 사용하므로 큐마다 SqsStrategy를 구현하는 구조임. 
  - ex. 현재 구조에서 CreateSqsStrategy의 type attribute 로는 CreateEventHandler와 SqsEventHandler 를 동시에 구현한 EventHandler를 사용함.
- resolver에 Event를 던지면 해당 Event를 처리하는 EventStrategy를 resolver에서 자동으로 찾아서 provide 하도록 함. 따라서 추후 큐가 확장되더라도 수정 범위가 크지 않음.


### 구조도 
<img width="1723" height="1007" alt="image" src="https://github.com/user-attachments/assets/1e238ad1-e5cd-4c8d-b0d9-121668e88532" />

- MessageBroker 는 MessageBrokerStrategy 에서 압축해서 사용 가능하지만, 인터페이스를 통한 타입 지정이 조금 더 명확하다고 생각해서 따로 둠.
- MessageBrokerStrategy 에서는 Map <String 타입 아이디, EventHandler> 으로 이벤트 핸들러 목록을 관리함. 또한 특정 이벤트의 처리 가능 여부를 반환하는 메서드가 존재해 resolver에서 요청받은 이벤트에 대해 알맞은 MessageBrokerStrategy 를 찾을 수 있음.
  - 살짝 찾아봤을 때는 Map <String 타입 아이디, EventHandler> 으로 이벤트 타입을 사용하는 구조가 Kafka 나 RabbitMQ 등의 다른 메세지 브로커에서도 공통적으로 사용 가능한 구조인 것 같아서 MessageBrokerStrategy 에서 관리하도록 함. 
  (다만 이 부분은 확실히 재사용이 가능한 건지 확신이 안서서... 아닌 것 같으면 알려주세요!)
- SqsStrartegy 는 Sqs 사용 시 필요한 queueUrl , 헤더 등을 관리하는 추상 클래스로, 큐를 동작에 따라 분리하고자하니 CreateSqsStrategy 등에서 사용하는 queueUrl 을 주입받고, 공통된 로직을 처리하는 구조. 
(즉, SqsStrartegy 의 구현체는 SQS 큐 단위로 존재하는 것 )
- CreateSqsStrategy 에서는 Type argument를 통해 SqsEventHandler 이며 CreateEventHandler 인 EventHandler들을 상속받은 MessageBrokerStrategy의 Map <String 타입 아이디, EventHandler> 에 주입함.
-  Event 및 EventHandler는 추상화 레이어를 조금 더 세세하게 나눠서 (ex. CreateEvent, CreateEventHandler) 타입 안정성을 높히고 알맞은 큐에서 사용하도록 함.


## ✨ Notification 
<!-- 참고할 사항을 작성해주세요. -->
- `application-cloud.yaml` 파일 수정해두었습니다

- Spring Cloud AWS SQS Starter 의존성 버전의 경우 기존 AWS 관련 의존성과 통일하여 3.1.1 을 사용하려 했으나 SqsListener의 JSON 직렬화가 간헐적으로 실패하는 이슈가 존재하여 최신 버전인 3.3.1을 선택했습니다.
- 지난번에 논의했던 SQS Container 를 찾아보니 메세지 폴링을 직접 On, Off 할 수 있어서 이런 세밀한 컨트롤이 필요한 상황에서 유용한 것 같습니다. 하지만 현재 저희는 그렇게 까지 세밀한 컨트롤이 필요한 상황은 아니라고 생각돼서 따로 사용하진 않았습니다.

jher235 committed:
CONFETI-28 feat: SQS provider, consumer 스레드 풀 설정
- memory bound 임을 고려하여 살짝 넉넉하게 잡음. 추후 트래픽을 예상하여 최적값 조율이 필요할 것으로 예상됨

jher235 committed:
CONFETI-28 feat: SQS Config 설정
- MESSAGE_PROVIDER_POOL 및 MESSAGE_CONSUMER_POOL 적용.

- 각 SqsListener 설정값은 스레드 풀을 제외하고 모두 비용 절감 및 효율성에 초점을 둠
- 최대 스레드 limit은 corePool과 동일한 30
-  한 번의 폴링 요청으로 수신할 수 있는 최대 메시지 수는 AWS SDK의 최대값인 10
- 폴링의 경우 최대 초인 20초로 설정.
- 메세지 삭제의 경우 성공 시 개별적으로 진행됨

jher235 committed:
CONFETI-28 feat: ArtistId 및 SongId로 객체의 존재 여부를 확인하는 메서드 추가

jher235 committed:
CONFETI-28 feat: Spring Cloud AWS SQS Starter 의존성 추가
- 기존 AWS 의존성과 버전을 3.1.1로 통일하려했으나 JSON 직렬화를 정상적으로 수행하지 못하는 경우가 존재하여 최신 버전인 3.3.1을 사용함

jher235 committed:
CONFETI-28 feat: MessageBroker 인터페이스 작성
- 처리할 Artist, Song 생성 이벤트를 파라미터로 받는 메서드 가짐

jher235 committed:
CONFETI-28 feat: Artist 생성 이벤트 클래스 작성
- toArtist 메서드의 경우 정적 팩토리보다 빌더패턴을 사용했는데 String, Integer 등의 타입이 공통적으로 사용되기 때문

jher235 committed:
CONFETI-28 feat: Song 생성 이벤트 클래스 작성
- toArtist 메서드의 경우 정적 팩토리보다 빌더패턴을 사용했는데 String, Integer 등의 타입이 공통적으로 사용되기 때문

jher235 committed:
CONFETI-28 feat: 에러메세지 추가

jher235 committed:
CONFETI-28 feat: Json 직렬화 및 역직렬화를 위한 Util 클래스 작성

jher235 committed:
CONFETI-28 feat: NotificationAgent 를 dev 환경에서 사용하기 위한 LocalNotificationAgent 작성

jher235 committed:
CONFETI-28 feat: ArtistService 작성

jher235 committed:
CONFETI-28 feat: SongService 작성

jher235 committed:
CONFETI-28 feat: MessageBroker의 Sqs 구현제인 SqsService 작성
- 헤더의 confetiEventType를 통해 이벤트 타입을 관리함
- 메시지 추적을 위해 traceId와 이벤트 타입을 속성(Attribute)에 포함함
- @SqsListener를 사용하여 큐의 메시지를 수신하며 eventHandlers 를 통해 이벤트별 핸들러 메서드를 호출하는 구조
- SQS 에 메세지를 provide 하는 로직은 비동기적이며 실패 시 재시도가 아니라 notify를 줌. Artist나 Song을 생성하는 이벤트 등록이 실패하더라도 치명적이지 않다고 판단했기 때문
- 각 이벤트 핸들러는 데이터베이스에 데이터를 생성하기 전, 해당 데이터의 존재 여부를 먼저 확인하여 멱등성을 보장함

jher235 committed:
Merge branch 'develop' into feat/CONFETI-28

# Conflicts:
#	build.gradle
#	src/main/java/org/sopt/confeti/global/message/ErrorMessage.java

jher235 committed:
CONFETI-28 feat: INTERNAL_SERVER_ERROR 로 통일 및 추가적인 ErrorMessage 삭제
- 애초에 기존 설계 방향이 서버 내부적으로 발생하는 예외를 자세하게 클라이언트에게 노출하지 않는 방향인 것 같아서 서버 내부 오류로 통일함

jher235 committed:
CONFETI-28 feat: Event 추상화
- 타입 아이디를 반환하는 default 메서드 작성

jher235 committed:
CONFETI-28 feat: event를 upper bound로 갖는 이벤트 핸들러 작성

jher235 committed:
CONFETI-28 feat: ArtistCreate, SongCreate 이벤트 핸들러 구현체 작성

jher235 committed:
CONFETI-28 feat: Event를 implement 하여 추상화 진행

jher235 committed:
CONFETI-28 refactor: SqsService에 EventHandler를 사용하여 추후 이벤트가 추가되더라도 코드를 건드릴 필요없이 확장 가능하도록 변경

jher235 committed:
CONFETI-28 refactor: 폴링 스레드 수를 2개로 변경

jher235 committed:
CONFETI-28 refactor: 각 handle 작업에서 DB 트랜잭션을 관리하도록 함

jher235 committed:
CONFETI-28 refactor: 폴링 스레드와 컨슈머 스레드 풀을 분리하여 비동기 작업을 수행하도록 변경

jher235 committed:
CONFETI-28 refactor: comsumer 스레드 풀의 사이즈 조정
- corepool은 20개. 2 (폴링 스레드 ) * 10 (한번에 수신하는 최대 메세지 수) 으로 계산했으며 DB 커넥션 풀 수보다 크지만, 작업 이후 message 삭제 Acknowledge 하는 부분 로직까지 수행하므로 조금 넉넉하게 설정함
- MaxPool 은 현재 사용될 일이 없지만, 추후 큐를 분리하는 등의 작업으로 스레드 풀 수가 부족할 경우를 대비하여 두 배인 40으로 설정함

ch1hyun commented on src/main/java/org/sopt/confeti/global/config/ThreadPoolConfig.java:
```diff
@@ -0,0 +1,40 @@
+
+@Configuration
+public class ThreadPoolConfig {
+
+    public static final String SQS_WORKER_PREFIX = "sqs-worker-";
+    public static final String MESSAGE_PROVIDER_POOL = "providerThreadPool";
+    public static final String MESSAGE_CONSUMER_POOL = "consumerThreadPool";
+
+    @Bean(MESSAGE_PROVIDER_POOL)
+    public ThreadPoolTaskExecutor messageProvider() {
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(20);
+        executor.setMaxPoolSize(40);
+        executor.setQueueCapacity(50);
+        executor.initialize();
+        return executor;
+    }
```
Provider에도 prefix를 붙이는건 어떨까요? 나중에 추적에도 용이할 것 같아요.

ch1hyun reviewed (changes requested):
나머진 이따 보겠습니다..!

jher235 committed:
CONFETI-28 refactor: sendEvent 메서드만 사용하도록 추상화 진행

jher235 committed:
CONFETI-28 feat: Event 패키지 수정

jher235 committed:
CONFETI-28 feat: CreateEvent 및 CreateEventHandler 작성
- 타입을 조금 더 구체적으로 사용하기 위함

jher235 committed:
CONFETI-28 feat: EventHandler default 메서드 추가 및 변경
- getSupportedTypeId 를 default 메서드로 사용하고 필요 시 오버라이드하도록 함. 인터페이스에 default 메서드를 사용하는 것이 좋은 패턴이 아닐 수 있어 고민됐지만, 필드가 필요하지 않은 현재 구조에서 모든 구현체들이 공통적으로 처리할 수 있는 부분을 default 메서드로 처리할 수 있다면 코드 중복을 줄일 수 있다고 판단함.

jher235 committed:
CONFETI-28 feat: strategy 및 resolver 를 적용하여 확장성을 높힘.
- SQS, Kafka 등의 MessageBroker 마다  MessageBrokerStrategy 를 구현해야 함.
- 현재는 SQS 내에서도 목적에 맞게 큐를 사용하므로 큐마다 SqsStrategy를 구현하는 구조임. 현재 구조에서 CreateSqsStrategy의 type attribute 로는 CreateEventHandler와 SqsEventHandler 를 동시에 구현한 EventHandler를 사용함

jher235 committed:
CONFETI-28 feat: maxMessagesPerPoll 을 2로 수정.
- 원래 구조를 maxConcurrentMessages 를 2로 두어 polling 하는 스레드를 2개로 두고 한번에 가져올 수 있는 작업 수를 10개로 둔 뒤 별도의 스레드 풀에서 이벤트 consume을 하려했지만, messagesPerPoll should be less than or equal to maxConcurrentMessages 오류가 발생하여 수정함.
- 추후 위에서 생각한 방식으로 메세지를 받기 위해 스케줄러를 따로 두는 구조를 구상 중

jher235 committed:
CONFETI-28 feat: messageProvider 스레드 풀에 prefix 추가

jher235 committed:
CONFETI-28 feat: Scheduler config 추가

jher235 committed:
CONFETI-28 feat: @SqsListener 대신 Scheduler 를 통해 직접 polling 하도록 변경
- 기존 @SqsListener 로는 한번에 가져오는 메세지 수가 polling 하는 스레드 수 보다 클 수 없었음. 그러나 현재 구조는 polling 스레드와 consume 하는 스레드가 다르므로 이를 세세하게 컨트롤하고자 직접 스케줄러를 통해 consume 하도록 변경함

jher235 committed:
CONFETI-28 feat: JsonUtil 을 JsonMapper 로 변경
- 기존에 JacksonConfig에서 설정해둔 ObjectMapper를 사용하기 위함

ch1hyun commented on src/main/java/org/sopt/confeti/global/messagebroker/MessageBrokerResolver.java:
```diff
@@ -0,0 +1,30 @@
+            });
+
+        strategy.sendEvent(event);
+    }
```
이거 항상 순회도는데, 빈 생성 시점에 Map으로 등록해두면 어떨까요?

jher235 replied:
처음 생각은 개수가 많아져도 10개 이상은 안갈테니 시간복잡도에 큰 영향이 없겠다고 생각해서 이렇게 구현했는데, Map을 두고 List를 사용한다고 해서 크게 코드가 깔끔해지거나 이해하기 쉬워지는 것 같지도 않아서 Map으로 변경하겠습니다~

ch1hyun commented on src/main/java/org/sopt/confeti/global/messagebroker/MessageBrokerStrategy.java:
```diff
@@ -0,0 +1,27 @@
+import lombok.Getter;
+import org.sopt.confeti.global.messagebroker.handler.EventHandler;
+import org.sopt.confeti.global.messagebroker.message.Event;
+
+@Getter
+public abstract class MessageBrokerStrategy implements MessageBroker {
+
+    private final Map<String, ? extends EventHandler<? extends Event>> eventHandlers;
+
+    protected MessageBrokerStrategy(List<? extends EventHandler<? extends Event>> eventHandlers) {
+        this.eventHandlers = eventHandlers.stream().collect(Collectors.toUnmodifiableMap(
+            EventHandler::getSupportedTypeId, Function.identity()));
+    }
+
+    public boolean isSupported(Class<? extends Event> eventClass) {
+        return eventHandlers.values()
+            .stream()
+            .anyMatch(eventHandler -> eventHandler.isSupported(eventClass));
+    }
```
찾을 때는 이벤트 클래스로 찾는데, 키를 문자열로 등록하신 이유가 있으실까요?
봤을 때 키를 사용하는 것 같지 않아서요!

jher235 replied:
처음에 메세지를 전송할 Strategy를 찾을 때는 이벤트 클래스를 사용하는데, 이벤트를 수실할 때는 TypeId를 String값으로 사용해서 해당 TypeId의 핸들러를 찾고 있습니다! 
-> 하나의 큐에서 여러 타입의 이벤트를 처리할 수 있어야 하다보니 이렇게 됐어요

ch1hyun commented on src/main/java/org/sopt/confeti/global/messagebroker/sqs/strategy/SqsStrategy.java:
```diff
@@ -0,0 +1,183 @@
+        Map<String, Object> messageAttributes = createMessageAttributes(event);
+        String serializedData = jsonMapper.toJson(event);
+
+        sqsTemplate.sendAsync(to -> to
+                .queue(queueName)
+                .payload(serializedData)
+                .headers(messageAttributes)
+            )
+            .exceptionally(exception -> {
+                String errorMessage = extractExceptionMessage(exception, event);
+                notificationAgent.notify(SlackNotificationType.HIGH_ERROR, errorMessage);
+                log.error("[SqsService send exception] {}", errorMessage, exception);
+                return null;
+            });
```
jw SqsAsyncClient가 아닌 SqsTemplate의 sendAsync를 통해 비동기 요청을 보낸 특별한 이유가 있나요?
보니까 deleteMessage에서만 클라이언트를 사용하는 것 같아서 차이점이 궁금해요

ch1hyun replied:
아 SqsTemplate 빈 등록 로직을 보니 SqsAsyncClient를 등록하는군요
그러면 deleteMessage는 직접 클라이언트를 통해서만 가능한거죠?

jher235 replied:
deleteMessage라는게 보낸 메세지를 consume 하기 전에 삭제하는 부분일까요?? 그렇다면 맞습니다~

ch1hyun commented on src/main/java/org/sopt/confeti/global/messagebroker/sqs/strategy/SqsStrategy.java:
```diff
@@ -0,0 +1,183 @@
+        return getEventHandlers()
+            .values()
+            .stream()
+            .filter(eventHandler -> eventHandler.isSupported(event.getClass()))
+            .findFirst()
+            .orElseThrow(() -> new ConfetiException(INTERNAL_SERVER_ERROR))
+            .getSupportedTypeId();
+    }
```
이것도 사전에 HashMap으로 등록하면 어떨까요?

jher235 replied:
넵 이전 Strategy List 부분과 동일한 이유로 이렇게 뒀었는데 반영하겠습니다~

ch1hyun commented on src/main/java/org/sopt/confeti/global/messagebroker/sqs/strategy/SqsStrategy.java:
```diff
@@ -0,0 +1,183 @@
+            .exceptionally(e -> {
+                log.error("[SQS Polling Exception] message: {}", receivedMessages, e);
+                return null;
+            });
```
여기서 10개의 메세지 중 하나의 메세지만 실패해도 나머지 성공한 9개의 메세지도 삭제가 안되는 것 같아요
확인 부탁드려요..!

jher235 replied:
앗 그렇네요 수정하겠습니다!!

ch1hyun commented on src/main/java/org/sopt/confeti/domain/applemusic/song/event/CreateSongEvent.java:
```diff
@@ -0,0 +1,25 @@
+import org.sopt.confeti.domain.applemusic.song.Song;
+import org.sopt.confeti.global.messagebroker.message.CreateEvent;
+
+public record CreateSongEvent(
```
이벤트를 sqs 패키지 내부에 두면 어떨까요?
이벤트 패키지와 도메인 패키지를 분리하면 좋을 것 같다는 생각이에요

ch1hyun commented on src/main/java/org/sopt/confeti/domain/applemusic/song/application/dto/request/CreateSongDTO.java:
```diff
@@ -0,0 +1,25 @@
+    Integer artworkHeight
+) implements CreateEvent {
+
+    public Song toSong() {
+        return Song.builder()
+            .songId(songId)
+            .name(name)
+            .artworkUrl(artworkUrl)
+            .artworkWidth(artworkWidth)
+            .artworkHeight(artworkHeight)
+            .build();
+    }
```
먼저 DTO로 변환한 뒤, Song 엔티티에서 직접 static 생성자를 통해 DTO를 받아 객체를 생성하는 흐름이 더욱 엔티티를 안전하게 사용한다고 생각해요
어떻게 생각하시나요!

jher235 replied:
저는 사실 이걸 이벤트이자 일종의 도메인 생성 Dto라고 생각했어서 패키지도 도메인 내부로 두었어요 (그래도 이벤트라는 개념에 더욱 가까우니 패키지는 옮기는게 더 좋을 것 같아서 반영하겠습니다~)
그렇다보니 해당 이벤트 객체 내에서 상위 개념의 Song 엔티티로 변환하는 로직을 갖고있는게 자연스럽다고 생각했어요. 그렇지만, 현재 상황처럼 Service 단 Dto를 따로 사용하고 있는 상황이라면 Event 를 Service 단 Dto로 변환하는 과정이 한번 더 있는게 좋겠네요! 

'Song 엔티티에서 직접 static 생성자를 통해 DTO를 받아 객체를 생성하는 흐름' 이라면 상위 레이어인 Song이 하위 레이어인 DTO를 알고 있어야 하는데, 이것보단 DTO가 엔티티를 알고 있는게 조금 더 자연스럽지 않을까요?? 
그래서 이걸 Event에서 toDTO 등의 메서드로 서비스단 Dto로 변경하고, 서비스단 Dto에서 다시 toSong 등의 메서드로 Song을 만들도록 하는 것은 어떨까요??

ch1hyun replied:
하위 레이어 -> 상위 레이어로 변환하는 흐름은 동의합니다!
다만, 도메인과 이벤트는 분리했으면 좋겠다는 생각이에요.
지금 작성하신 CreateSongDTO는 서비스 레이어에서 Song 엔티티를 생성하기 위해 데이터를 전달할 객체이고, 도메인 모델이라고 생각합니다.

서비스 레이어의 Song 엔티티 생성 엔드포인트에 접근하는 주체가 이벤트일 수도 있고, 비동기 콜백일 수도 있고, 동기 처리일 수도 있다고 생각해요. 꼭 이벤트일 보장이 없는 것 같습니다.
그래서 Sqs 이벤트 처리와 이벤트 객체는 이벤트 패키지로 분리하고 도메인은 도메인에 넣는 방향을 제안드린 거였습니다!

ch1hyun replied:
여튼 지금 구조는 제 의견이 반영되어있는 상태인 것 같네요!

jher235 replied:
넵 맞아요!! 저도 서비스단에서 엔티티를 생성 시에는 공통적인 DTO를 사용하는 것이 더 좋다는 의견에는 100% 동의해서 반영해두었습니다~

ch1hyun reviewed (changes requested):
고생하셨습니다!!!
추상화가 잘 된것 같아요
너무너무 좋습니다~~

jher235 committed:
CONFETI-28 refactor: 내부 핸들러의 구조를 List에서 Map으로 변경
- 기존에는 개수가 많아져도 10개 이상은 안갈테니 시간복잡도에 큰 영향이 없겠다고 생각해서 이렇게 구현했는데, Map을 두고 List를 사용한다고 해서 크게 코드가 깔끔해지거나 이해하기 쉬워지는 것 같지도 아니라고 판단해서 Map으로 변경
- handleEvent 의 파라미터 네이밍을 type에서 typeId로 변경

jher235 committed:
CONFETI-28 refactor: Event 패키지 구조 수정
- 도메인 하위에서 messagebroker 하위로 변경. dto 보다는 이벤트의 성격이 더 강하므로

jher235 committed:
CONFETI-28 feat: 메세지 처리 중 예외 발생 시 다른 메세지 처리 로직에 전파되지 않도록 로직 수정
- 기존 processMessage 에서 단일 메세지 처리 후 직접 deleteMessage까지 진행하도록 함

jher235 committed:
CONFETI-28 refactor: 폴링 관련 상수화
- 받아오는 최대 메세지 개수 및 타임아웃 시간을 상수 처리

jher235 committed:
CONFETI-28 refactor: 엔티티 생성 시 CreateDto를 사용하도록 변경
- Event -> CreateDto -> Entity 플로우로 엔티티를 생성하도록 함.
- 상위 레이어인 Entity는 하위 레이어인 Dto를 모르고, Dto는 또 자신의 하위 레이어인 Event를 모르도록 설계

ch1hyun commented:
@claude 위 컨텍스트를 모두 잊고 다시 리뷰해줘
